### DISCUSSION TITLE
Backpropagating through a circuit

### DISCUSSION THREAD

**Discussion by BrunoLiegiBastonLiegi** [qiboteam/qibo#726] [Q&A] [answered]
Hi, I was wondering whether it were possible to backpropagate through a circuit, @MatteoRobbiati, maybe you could help me figure this out.
In detail, I want to generate several different circuits by sampling their gates according to a distribution that depends on some parameters. For the moment, let's say that all the gates I sample don't depend on any parameter or that their parameters are fixed. Each sampled circuit is then run and I calculate the expected value of an observable (always the same one) at the end of it. If I define a loss based on the expected values I get, will I be able to backpropagate the loss, through the circuits, to the distribution parameters? In other words, will I be able to train for the optimal distribution parameters that minimize my loss?
What if I unfreeze the gates' parameters instead? I would have the composition of the gradient coming from the gates and the one coming from the distribution, right?

**Comment by scarrazza** [comment]
Just a quick remark, probably unrelated, if you use the tensorflow backend you should be able to get the analytic expression for all derivatives. Obviously the PSR would be better, because you can run on hardware.

**Comment by MatteoRobbiati** [comment, marked as answer]
I was thinking about the same thing @scarrazza says. 
Your problem seems to be an hyper-optimization one. In this case you are not directly derivating with respect to the gates' parameters, but you are searching for an instrument which allows you to calculate the derivative of the circuit's output with respect, e.g. the number of layers , right?

Tensorflow's `GradientTape` module is able to track all the state vector's modification under your assumptions (you must define all your elements inside the Gradient Tape). 

I must have a look deeply in literature searching for some hardware-compatible hyper-optimization techinques (of course you can try with Genetic Algos, but I am not sure it will be worth).

Do you want to have a meeting for discussing this?

**Reply by BrunoLiegiBastonLiegi** [reply]
Yes, precisely, in the first case the gates are not parametrized (or at least they are frozen). I'll take a look at ``GradienTape`` then. The thing that is not clear to me though, is whether with the tensorflow backend you are still able to run the circuit on hardware.
Anyway, yes thank you I would love to have a meeting!

**Reply by MatteoRobbiati** [reply]
You cannot run the gradient tape optimization on the hardware using tensorflow because (as you said) it use back-propagation.
Unfortunately the B-P is impossible to be performed on a circuit (because we cannot know the expectation value over the state we have in the middle of the run). We can split the discussion into two branches:

- if you wonder about the derivative of the circuit with respect to the variational parameters, let's use the parameter shift rule;
- if you want to do hyper-optimization let's use some proper algorithm (such as TPE, random search and so on).

And talking about the specific topic of the derivative of the circuit's family with respect to the hyper-parameters: we need to think deeply inside the topic.

Let's send me an e-mail and we can arrange a meeting :)

**Reply by BrunoLiegiBastonLiegi** [reply]
Thanks, I think you are right, probably it's better to go with some genetic or hyperoptimization algorithm, such as CMA or TPE.
